pkg/pillar/hardware/smart.go

```diff
@@ -0,0 +1,227 @@
+	if err := json.Unmarshal(data, &disks); err != nil {
+		return nil, fmt.Errorf("error while parsing SMART data. file:%s err:%s", pathDisksList, err)
+	}
+
+	if disks.Smartctl.ExitStatus != 0 {
+		return nil, fmt.Errorf("error-code: %d, info:%v",
+			disks.Smartctl.ExitStatus,
+			disks.Smartctl.Messages)
+	}
+
+	return disks, nil
+}
+
+func getAttrFromJSON(attrList types.SmartAttributes, id int) *info.SmartAttr {
+	for _, data := range attrList.Table {
+		if data.ID == id {
+			attr := new(info.SmartAttr)
+			attr.Id = uint32(data.ID)
+			attr.Value = uint64(data.Value)
+			attr.Worst = uint64(data.Worst)
+			attr.Thresh = uint64(data.Thresh)
+			attr.WhenFailed = data.WhenFailed
+			attr.RawValue = uint64(data.Raw.Value)
+			return attr
+		}
+	}
+	return nil
+}
+
+func getSmartAttr(attrList types.SmartAttributes) *info.SmartMetric {
+	smart := new(info.SmartMetric)
+	smart.CurrentPendingSector = getAttrFromJSON(attrList, idCurrentPendingSector)
+	smart.PowerCycleCount = getAttrFromJSON(attrList, idPowerCycleCount)
+	smart.PowerOnHours = getAttrFromJSON(attrList, idPowerOnHours)
+	smart.ReallocatedSectorCt = getAttrFromJSON(attrList, idReallocatedSectorCt)
+	return smart
+}
+
+// ReadDiskSmartInfoFromJSON read JSON file and return *types.DeviceSmartInfo
+func ReadDiskSmartInfoFromJSON(diskName string) (*types.DeviceSmartInfo, error) {
+	path := getPathForFileDisk(diskName)
+	data, err := readFWithMaxSize(path, maxSmartCtlSize)
+	if err != nil {
+		return nil, fmt.Errorf("read file with SMART info for %s disk failed. %s", path, err)
+	}
+
+	deviceInfo := new(types.DeviceSmartInfo)
+	if err := json.Unmarshal(data, &deviceInfo); err != nil {
+		return nil, fmt.Errorf("error while parsing SMART data. %s", err)
+	}
+
+	if deviceInfo.Smartctl.ExitStatus != 0 {
+		return nil, fmt.Errorf("error-code: %d, info:%v",
+			deviceInfo.Smartctl.ExitStatus,
+			deviceInfo.Smartctl.Messages)
+	}
+
+	return deviceInfo, nil
+}
+
+// getAndComparisonSmartDataForDisk takes a disk name as input, then tries
+// to collect old data and get new data, compares if necessary and
+// always returns new data in the form of:
+//
+// *types.DeviceSmartInfo - regardless of the result, we always send last new data,
+// except for situations with errors when receiving this data
+//
+// bool - Need to update information, true - necessary, false - data is the same
+//
+// error - Errors with getting new data, if any
+func getAndComparisonSmartDataForDisk(diskName string) (*types.DeviceSmartInfo, bool, error) {
+	needUpdate := false
+	oldData, err := ReadDiskSmartInfoFromJSON(diskName)
+	if err != nil {
+		// it's normal situation
+		needUpdate = true
+	}
+
+	// Get new S.M.A.R.T data for disks
+	args := []string{"--all", diskName, "--json"}
+	pathFile := getPathForFileDisk(diskName)
+	if err := smartctlExec(pathFile, args...); err != nil {
+		return nil, false, fmt.Errorf("get disk info from smartctl filed. err: %s", err)
+	}
+
+	newData, err := ReadDiskSmartInfoFromJSON(diskName)
```

Comment:
Reading this it is far from obvious that it would ever get different data from the second invocation of ReadDiskSmartInfoFromJSON (have to look at all of the associated functions to find the fact that things are cached in a file).

How about introducing a function called UpdateDiskSmartInfoCache (which does line 177-182) and rename ReadDiskSmartInfoFromJSON to ReadDiskSmartInfoFromCache()? Those names would give a lot more clues.